File: smokers/java/Smokers.java
import java.util.Random;
import java.util.concurrent.Semaphore;

public class Smokers{
 public static void main(String[] args) {
    Pushers pushers = new Pushers();

    new AgentThread(pushers);

    new PusherAThread(pushers);
    new PusherBThread(pushers);
    new PusherCThread(pushers);

    new TobaccoSmoker(pushers);
    new PaperSmoker(pushers);
    new MatchSmoker(pushers);
 }
}

class AgentThread implements Runnable { 
    Pushers pushers;

    AgentThread(Pushers pushers) {
        this.pushers = pushers;

        new Thread(this, "AgentThread").start();;
    }

    public void run() {
        while (true) {
            this.pushers.agent();
        }
    }
}

class PusherAThread implements Runnable {
    Pushers pushers;

    PusherAThread(Pushers pushers) {
        this.pushers = pushers;

        new Thread(this, "PusherAThread").start();;
    }

    public void run() {
        while (true) {
            this.pushers.pusherA();
        }
    }
}

class PusherBThread implements Runnable {
    Pushers pushers;

    PusherBThread(Pushers pushers) {
        this.pushers = pushers;

        new Thread(this, "PusherBThread").start();;
    }

    public void run() {
        while (true) {
            this.pushers.pusherB();
        }
    } 
}

class PusherCThread implements Runnable {
    Pushers pushers;

    PusherCThread(Pushers pushers) {
        this.pushers = pushers;

        new Thread(this, "PusherCThread").start();;
    }

    public void run() {
        while (true) {
            this.pushers.pusherC();
        }
    } 
}

class TobaccoSmoker implements Runnable {
    Pushers pushers;

    TobaccoSmoker(Pushers pushers) {
        this.pushers = pushers;

        new Thread(this, "TobaccoSmoker").start();;
    }

    public void run() {
        try{
            while (true) {
                this.pushers.smokerT();
            }
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        }
    } 
}

class PaperSmoker implements Runnable {
    Pushers pushers;

    PaperSmoker(Pushers pushers) {
        this.pushers = pushers;

        new Thread(this, "PaperSmoker").start();;
    }

    public void run() {
        try {
            while (true) {
                this.pushers.smokerP();
            }
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        }
    } 
}

class MatchSmoker implements Runnable {
    Pushers pushers;

    MatchSmoker(Pushers pushers) {
        this.pushers = pushers;

        new Thread(this, "MatchSmoker").start();;
    }

    public void run() {
        try {
            while (true) {
                this.pushers.smokerM();
            }
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        }
    } 
}

class Pushers {
    int tobaccoOnTable = 0;
    int paperOnTable = 0;
    int matchOnTable = 0;

    static Semaphore mutex = new Semaphore(1);
    static Semaphore agentSem = new Semaphore(1);
    
    static Semaphore tobacco = new Semaphore(0);
    static Semaphore paper = new Semaphore(0);
    static Semaphore match = new Semaphore(0);

    static Semaphore tobaccoSmoker = new Semaphore(0);
    static Semaphore paperSmoker = new Semaphore(0);
    static Semaphore matchSmoker = new Semaphore(0);    

    void pusherA() {
        try {
            tobacco.acquire();
            mutex.acquire();
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        } 
        if (paperOnTable > 0) {
            paperOnTable--;
            matchSmoker.release();
        } else if (matchOnTable > 0) {
            matchOnTable--;
            paperSmoker.release();
        } else {
            tobaccoOnTable++;
        }
        mutex.release();
    }

    void pusherB() {
        try {
            paper.acquire();
            mutex.acquire();
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        }
        if (tobaccoOnTable > 0) {
            tobaccoOnTable--;
            matchSmoker.release();
        } else if (matchOnTable > 0) {
            matchOnTable--;
            tobaccoSmoker.release();
        } else {
            paperOnTable++;
        }
        mutex.release();
    }

    void pusherC() {
        try {
            match.acquire();
            mutex.acquire();
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        }
        if (tobaccoOnTable > 0) {
            tobaccoOnTable--;
            paperSmoker.release();
        } else if (paperOnTable > 0) {
            paperOnTable--;
            tobaccoSmoker.release();
        } else {
            matchOnTable++;
        }
        mutex.release();
    }

    void agent() {
        try {
            agentSem.acquire();
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        }
        Random rand = new Random();
        int dice = rand.nextInt(3);

        switch (dice) {
            case 0:
                tobacco.release();
                paper.release();
                break;

            case 1:
                paper.release();
                match.release();
                break;

            case 2:
                tobacco.release();
                match.release();
                break;
        }
    }

    void smokerT() throws InterruptedException {
        tobaccoSmoker.acquire();
        makeCigarettes("TobaccoHolder");
        agentSem.release();
        smoke("TobaccoHolder");
    }

    void smokerP() throws InterruptedException {
        paperSmoker.acquire();
        makeCigarettes("PaperHolder");
        agentSem.release();
        smoke("PaperHolder");
    }

    void smokerM() throws InterruptedException {
        matchSmoker.acquire();
        makeCigarettes("MatchHolder");
        agentSem.release();
        smoke("MatchHolder");
    }

    void makeCigarettes(String holderId) throws InterruptedException {
        System.out.println(holderId + " is making cigarettes.");
        Thread.sleep(new Random().nextInt(1001));
    }

    void smoke(String holderId) throws InterruptedException {
        System.out.println(holderId + " is smoking.");
        Thread.sleep(new Random().nextInt(1001));
    }
}

File: smokers/python/smokers.py
import threading
import random
import time
import psutil

class Pushers():
    mutex = threading.Semaphore(1)
    agentSem = threading.Semaphore(1)

    tobacco = threading.Semaphore(0)
    paper = threading.Semaphore(0)
    match = threading.Semaphore(0)

    tobaccoSmoker = threading.Semaphore(0)
    paperSmoker = threading.Semaphore(0)
    matchSmoker = threading.Semaphore(0)

    tobaccoOnTable = 0
    paperOnTable = 0
    matchOnTable = 0

    def agent(self):
        while 1:
            self.agentSem.acquire()
            randomInt = random.randint(0, 2)
            if (randomInt == 0):
                self.tobacco.release()
                self.paper.release()
            elif (randomInt == 1):
                self.paper.release()
                self.match.release()
            else:
                self.tobacco.release()
                self.match.release()
            print "CPU: ", psutil.cpu_percent()

    def PusherA(self):
        while 1:
            self.tobacco.acquire()
            self.mutex.acquire()
            if self.paperOnTable:
                self.paperOnTable -= 1
                self.matchSmoker.release()
            elif self.matchOnTable:
                self.matchOnTable -= 1
                self.paperSmoker.release()
            else:
                self.tobaccoOnTable += 1
            self.mutex.release()
	
    def PusherB(self):
        while 1:
            self.paper.acquire()
            self.mutex.acquire()
            if self.tobaccoOnTable:
                self.tobaccoOnTable -= 1
                self.matchSmoker.release()
            elif self.matchOnTable:
                self.matchOnTable -= 1
                self.tobaccoSmoker.release()
            else:
                self.paperOnTable += 1
            self.mutex.release()
        
    def PusherC(self):
        while 1:
            self.match.acquire()
            self.mutex.acquire()
            if self.tobaccoOnTable:
                self.tobaccoOnTable -= 1
                self.paperSmoker.release()
            elif self.paperOnTable:
                self.paperOnTable -= 1
                self.tobaccoSmoker.release()
            else:
                self.matchOnTable += 1
            self.mutex.release()

    def TobaccoHoldingSmoker(self):
        while 1:
            self.tobaccoSmoker.acquire()
            self.makeCigarettes("TobaccoHolder")
            self.agentSem.release()
            self.smoke("TobaccoHolder")

    def PaperHoldingSmoker(self):
        while 1:
            self.paperSmoker.acquire()
            self.makeCigarettes("PaperHolder")
            self.agentSem.release()
            self.smoke("PaperHolder")

    def MatchHoldingSmoker(self):
        while 1:
            self.matchSmoker.acquire()
            self.makeCigarettes("MatchHolder")
            self.agentSem.release()
            self.smoke("MatchHolder")
        
    def makeCigarettes(self, holderId):
        print holderId, " is makinhg cigarettes."
        time.sleep(0.1)
        
    def smoke(self, holderId):
        print holderId, " is smoking."
        time.sleep(0.1)

def main():
    pushers = Pushers()

    pusherA = threading.Thread(target=pushers.PusherA)
    pusherB = threading.Thread(target=pushers.PusherB)
    pusherC = threading.Thread(target=pushers.PusherC)
    agent = threading.Thread(target=pushers.agent)

    tobaccoSmoker = threading.Thread(target=pushers.TobaccoHoldingSmoker)
    paperSmoker = threading.Thread(target=pushers.PaperHoldingSmoker)
    matchSmoker = threading.Thread(target=pushers.MatchHoldingSmoker)

    pusherA.start()
    pusherB.start()
    pusherC.start()
    agent.start()

    tobaccoSmoker.start()
    paperSmoker.start()
    matchSmoker.start()

main()
	
